node/network/statement-distribution/src/lib.rs

```diff
@@ -0,0 +1,1376 @@
+
+const BENEFIT_VALID_STATEMENT: Rep = Rep::new(5, "Peer provided a valid statement");
+const BENEFIT_VALID_STATEMENT_FIRST: Rep = Rep::new(
+	25,
+	"Peer was the first to provide a valid statement",
+);
+
+/// The maximum amount of candidates each validator is allowed to second at any relay-parent.
+/// Short for "Validator Candidate Threshold".
+///
+/// This is the amount of candidates we keep per validator at any relay-parent.
+/// Typically we will only keep 1, but when a validator equivocates we will need to track 2.
+const VC_THRESHOLD: usize = 2;
+
+/// The statement distribution subsystem.
+pub struct StatementDistribution;
+
+impl<C> Subsystem<C> for StatementDistribution
+	where C: SubsystemContext<Message=StatementDistributionMessage>
+{
+	fn start(self, ctx: C) -> SpawnedSubsystem {
+		// Swallow error because failure is fatal to the node and we log with more precision
+		// within `run`.
+		SpawnedSubsystem(run(ctx).map(|_| ()).boxed())
+	}
+}
+
+fn network_update_message(n: NetworkBridgeEvent) -> AllMessages {
+	AllMessages::StatementDistribution(StatementDistributionMessage::NetworkBridgeUpdate(n))
+}
+
+/// Tracks our impression of a single peer's view of the candidates a validator has seconded
+/// for a given relay-parent.
+///
+/// It is expected to receive at most `VC_THRESHOLD` from us and be aware of at most `VC_THRESHOLD`
+/// via other means.
+#[derive(Default)]
+struct VcPerPeerTracker {
+	local_observed: arrayvec::ArrayVec<[Hash; VC_THRESHOLD]>,
+	remote_observed: arrayvec::ArrayVec<[Hash; VC_THRESHOLD]>,
+}
+
+impl VcPerPeerTracker {
+	// Note that the remote should now be aware that a validator has seconded a given candidate (by hash)
+	// based on a message that we have sent it from our local pool.
+	fn note_local(&mut self, h: Hash) {
+		if !note_hash(&mut self.local_observed, h) {
+			log::warn!("Statement distribution is erroneously attempting to distribute more \
+				than {} candidate(s) per validator index. Ignoring", VC_THRESHOLD);
+		}
+	}
+
+	// Note that the remote should now be aware that a validator has seconded a given candidate (by hash)
+	// based on a message that it has sent us.
+	//
+	// Returns `true` if the peer was allowed to send us such a message, `false` otherwise.
+	fn note_remote(&mut self, h: Hash) -> bool {
+		note_hash(&mut self.remote_observed, h)
+	}
+}
+
+fn note_hash(
+	observed: &mut arrayvec::ArrayVec<[Hash; VC_THRESHOLD]>,
+	h: Hash,
+) -> bool {
+	if observed.contains(&h) { return true; }
+
+	if observed.is_full() {
+		false
+	} else {
+		observed.try_push(h).expect("length of storage guarded above; \
+			only panics if length exceeds capacity; qed");
+
+		true
+	}
+}
+
+// knowledge that a peer has about goings-on in a relay parent.
+#[derive(Default)]
+struct PeerRelayParentKnowledge {
+	// candidates that the peer is aware of. This indicates that we can
+	// send other statements pertaining to that candidate.
+	known_candidates: HashSet<Hash>,
```

Comment:
isn't this superfluous given `received_message_count`

Reply:
No, because the `known_candidates` also encompass sent messages, which leave no record in the `received_message_count`.